src/infra/core/src/transform_service_impl.rs

```diff
@@ -465,16 +474,31 @@ impl TransformServiceImpl {
         let schema_slice = if let Some(h) = data_slices.last() {
             h.clone()
         } else {
-            // TODO: This will not work with schema evolution
+            type Flag = MetadataEventTypeFlags;
+            type Decision = MetadataVisitorDecision;
+
             input_chain
-                .iter_blocks()
-                .filter_data_stream_blocks()
-                .filter_map_ok(|(_, b)| b.event.new_data)
-                .try_first()
-                .await
-                .int_err()?
-                .unwrap() // Already checked that none of the inputs are empty
-                .physical_hash
+                // TODO: This will not work with schema evolution
+                .reduce(
+                    None,
+                    Decision::NextOfType(Flag::DATA_BLOCK),
+                    |state, _, block| {
+                        let Some(data_block) = block.as_data_stream_block() else {
+                            unreachable!()
+                        };
+
+                        let Some(new_data) = data_block.event.new_data else {
```

Comment:
I don't think this change improves anything. We might need a special skip list that skips over watermark-only data blocks, but this code will not be able to utilize it anyway, so might be better to leave the old simpler version.

Reply:
I think you might like the new version better -- Done